src/btree/bt_cursor.c

```diff
@@ -778,6 +833,38 @@ __wt_btcur_search(WT_CURSOR_BTREE *cbt)
     return (ret);
 }
 
+/*
+ * __btcur_search_near_bounds_reposition --
+ *     This function validates whether a given key is within the provided cursor bounds. If not the
+ *     search near key is updated to the nearest bound.
+ */
+static int
+__btcur_search_near_bounds_reposition(WT_SESSION_IMPL *session, WT_CURSOR_BTREE *cbt)
+{
+    WT_BTREE *btree;
+    WT_CURSOR *cursor;
+    bool key_out_of_bounds, upper;
+
+    btree = CUR2BT(cbt);
+    cursor = &cbt->iface;
+    key_out_of_bounds = upper = false;
+
+    WT_ASSERT(session, WT_CURSOR_HAS_BOUNDS(cursor));
+    if (btree->type != BTREE_ROW)
+        WT_RET(ENOTSUP);
+
+    /*
+     * We need to assume that the bounds are inclusive here, suppose a caller calls with the search
+     * key set to the lower bound but also specifies that the lower bound isn't inclusive. We cannot
+     * know which key to set the lower bound to so we set it to the lower bound. The same is true
+     * for the upper bound check.
```

Comment:
I don't understand this.  Suppose bounds are 10 and 20, and cursor->key is 20, and cursor is marked as not inclusive.  I _think_ this code will not say the key is out of bounds, and so the key is not modified.
However, in the same situation, suppose I called `key_within_bounds` with `force_inclusive` set to false.  It would return out of bounds, and I'd set the key to 10.  So no difference.  All to say, does the `force_inclusive` do anything useful?  Am I missing some situation?

Reply:
In scenario one where the bound are 10 and 20, and the key is 20 and not inclusive. Assuming we don't force it to be inclusive we would then decide the key is not within the bounds and then have to set the key to the nearest key to the bound. However without complex search logic we cannot know what key is nearest to the upper bound key of 20(not inclusive) so to avoid an additional set_key operation we just assume the key is inclusive so we don't adjust the search key. The expectation is that the subsequent call to cursor_valid will return false due to being on an out of bounds key 20. 

If we did set the key to 10 we could wrongly return 10, or 11 for example. But the nearest record in this case would be 19.

Reply:
@luke-pearson Whoops, wrote my comment too quickly, the scenario meant to say `cursor->key` is 10.  So again:

Suppose bounds are 10 and 20, and `cursor->key` is 10, and cursor is marked as not inclusive. I think this code will not say the key is out of bounds, and so the key is not modified (remains 10).

Now, suppose the code is different and the call to `key_within_bounds` looks like this:
```
WT_RET(__key_within_bounds(session, cursor, &cursor->key, false, &key_out_of_bounds, &upper));
```
(That is, changed `true` to `false`).  It returns out of bounds, and the code sets the key to 10.  So there's no difference.  I'm just wondering if we even need that `force_inclusive` argument at all, I think it makes the code a little harder to understand.  Or am I missing a different purpose?

Reply:
I just thought the additional call to set_key wasn't very performant if we want to do that or code clarity then we can remove the force inclusive. Essentially on non-inclusive bounds with search keys that match the bound it becomes a sort of no-op.

Reply:
Okay, performance.  It's possible, with a large key in this particular situation it may make a difference.  I leave it to you to decide.  I'll take a look at the python tests tomorrow.

Reply:
Jie raised a similar point so I think i'll remove it at this stage, if we find it is a performance issue we can always add it back.